### ISSUE TITLE
kv: mass INSERTs easily run into TransactionRetryWithProtoRefreshError

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#98553]
Initially found [using Striim](https://docs.google.com/document/d/17-lUS-_Syw5kyu-LhlWxYaBqbgFxVFQnkZAcCEvz7q0/edit#).
Related COPY issue: https://github.com/cockroachdb/cockroach/issues/90656
May also go away with: https://github.com/cockroachdb/cockroach/issues/91831

When using Striim using the default settings, we usually run into `TransactionRetryWithProtoRefreshError` (`RETRY_ASYNC_WRITE_FAILURE`).

This is usually at about the 5MM row mark where rows are inserted one at a time on a single thread as follows (using the pg wire protocol):
```
Parse <stmt>

Bind <stmt> <args>
Execute

# ...
# Bind/Execute 255 times in total
#

Sync
```

This is a very common JDBC pattern.

Whilst we do have a workaround for now, we think it is beneficial to make the database more tolerant of these kinds of issues.

Jira issue: CRDB-25326

gz#16766

gz#17451

**Comment by nvanbenschoten** [comment]
> This is usually at about the 5MM row mark where rows are inserted one at a time on a single thread as follows

Do we know why rows are being inserted one at a time? Are these tools not batching the insertion of multiple rows? I'm asking because:
1. batching is an important throughput optimization
2. KV doesn't pipeline batched inserts, so they're not subject to these `RETRY_ASYNC_WRITE_FAILURE` failures

**Comment by otan** [comment]
We actually have a workaround for batching these now and these errors appear less / not at all. So from our side, we're unblocked for now. Combed with @cucaroach 's future work on batching INSERTs in the same Sync request, this may be a "non-issue".

(Opened for now as tracking inter-team dependencies for the next planning cycle by request from my PM/EM but we can probably get away from not doing anything about this.)

**Comment by cucaroach** [comment]
> Do we know why rows are being inserted one at a time? Are these tools not batching the insertion of multiple rows? I'm asking because:
> 
> 1. batching is an important throughput optimization
> 2. KV doesn't pipeline batched inserts, so they're not subject to these `RETRY_ASYNC_WRITE_FAILURE` failures

This pattern of "Bind <stmt> <args>,Execute" is how the postgres driver implements prepared batches [1].  The problem is we execute each one when we see the Execute instead of justing pushing it onto a queue and waiting until the Sync comes.

[1] The defacto/correct way to do bulk inserts in my experience, most other DBMS have moved on from string concatentation for bulk operations but the Go/PG/CRDB community seems happy with concatenating values clauses with commas.

**Comment by cucaroach** [comment]
These batch prepared insert statements also can be much larger than 255 (like 10k rows) and like COPY the batch is implicitly considered an auto commit boundary, ie the entire batch is atomically committed.   Like COPY I think we will probably explore buffering these types of inserts because applications aren't typically coded to handle retries in this situation (buffering is hard), like DMS they typically just declare failure and force the higher level operation to be repeated possibly involving manual intervention.  

Anyways I think part of the ask here is that common errors on large batches be scrutinized by the KV team and we think about ways to avoid them.   The thinking is if the errors are rare enough maybe we don't need the SQL layer buffering and we can avoid the memory pressure hazards the buffer everything solution carries with it (we're already thinking we should probably entertain spilling to disk as well).

As an aside the "if the world was my oyster" solution would be if inserts could be done with a savepoint/nested transaction model where if a retriable error occurred inserting a batch we could JUST retry that batch and everything that had been inserted successfully so far was still in a ready to be committed state.   I realize that's probably not a small ask.

**Comment by nvanbenschoten** [comment]
> Like COPY I think we will probably explore buffering these types of inserts because applications aren't typically coded to handle retries in this situation

This is great to avoid these errors. I want to make sure my second point above wasn't missed though, because it's another strong reason to do what you're suggesting and buffer and then combine these inserts:

These errors are only possible if the transaction is pipelining writes. We don't pipeline writes for INSERT statements that write more than 128 keys because there's a throughput/latency trade-off as the KV batch size grows. So the fact that we're seeing these errors indicates that we're performing many small INSERTs. Doing so leaves a significant amount of throughput on the table because it prevents batching at the RPC level, the KV evaluation level, the Raft level, or the disk level. Doing so and also disabling pipelining (is that your current workaround?) is going to be even worse because we'd then be performing the consensus rounds and disk writes of each small INSERT sequentially.

We can look into reducing the rate at which you hit these errors, but there's still the hidden issue that the ingestion is going to be slow if it behaves like this. Maybe that's all fine and we don't care about the performance of these workloads. What do you think?

**Comment by cucaroach** [comment]
Prepared inserts right now are executed row by row so yeah they will frequently be much fewer than 128 keys.   My intent is to apply the same batching and encoding logic that we're using for COPY in 23.1 to prepared batch inserts in 23.2 in which case KV batches will be much larger.   I suspect doing so will result in multiple orders of magnitude speedup for all the reasons you alluded too.  

I do think we care about this workload.  My personal suspicion is that we've probably had prospective customers who's apps require fast bulk inserts who have tried JDBC style prepared batch inserts against CRDB and after seeing the woeful results promptly uninstalled it and we never heard from them.  I don't think STRIIM is unique in relying on this approach for fast bulk inserts.
